kore/src/Kore/Builtin/Set.hs

```diff
@@ -270,66 +330,99 @@ evalUnit =
   where
     evalUnit0 _ resultSort =
         \case
-            [] -> returnSet resultSort Set.empty
+            [] -> returnConcreteSet resultSort Set.empty
             _ -> Builtin.wrongArity unitKey
 
 evalConcat :: Builtin.Function
 evalConcat =
     Builtin.functionEvaluator evalConcat0
   where
-    ctx = concatKey
-    evalConcat0 :: Builtin.FunctionImplementation
-    evalConcat0 _ resultSort = \arguments ->
-        Builtin.getAttemptedAxiom $ do
-            let (_set1, _set2) =
-                    case arguments of
-                        [_set1, _set2] -> (_set1, _set2)
-                        _ -> Builtin.wrongArity concatKey
-                leftIdentity = do
-                    _set1 <- expectBuiltinSet ctx _set1
-                    if Set.null _set1
-                        then
-                            Builtin.appliedFunction
-                            $ Pattern.fromTermLike _set2
-                        else empty
-                rightIdentity = do
-                    _set2 <- expectBuiltinSet ctx _set2
-                    if Set.null _set2
-                        then
-                            Builtin.appliedFunction
-                            $ Pattern.fromTermLike _set1
-                        else empty
-                bothConcrete = do
-                    _set1 <- expectBuiltinSet ctx _set1
-                    _set2 <- expectBuiltinSet ctx _set2
-                    if Set.null (Set.intersection _set1 _set2)
-                        then returnSet resultSort (_set1 <> _set2)
-                        else return (AttemptedAxiom.Applied mempty)
-            leftIdentity <|> rightIdentity <|> bothConcrete
+    evalConcat0
+        :: forall variable m
+        .  (Ord variable, SortedVariable variable)
+        => MonadSimplify m
+        => TermLikeSimplifier
+        -> Sort
+        -> [TermLike variable]
+        -> m (AttemptedAxiom variable)
+    evalConcat0 _ resultSort arguments = Builtin.getAttemptedAxiom $ do
+        tools <- askMetadataTools
+
+        let (_set1, _set2) =
+                case arguments of
+                    [_set1, _set2] -> (_set1, _set2)
+                    _ -> Builtin.wrongArity concatKey
+
+            emptyResult :: AttemptedAxiom variable
+            emptyResult = AttemptedAxiom.Applied mempty
+
+        case (toNormalizedSet tools _set1, toNormalizedSet tools _set2) of
+            (Bottom, _) -> return emptyResult
+            (_, Bottom) -> return emptyResult
+            -- The NormalizedSet matching is useful only for getting
+            -- notified when new fields are being added.
+            ( Normalized _set1@(Domain.NormalizedSet _ _ _)
+              , Normalized _set2
```

Comment:
Fixed. I wonder if we should allow this kind of half-indentation for `case` matchers that span more than one line. Maybe not worth the trouble, since it works nicely only for tuples and lists.